File: apps/rn/src/domain/rede.test.ts
import { PERGUNTA_FALHA, estimarComFalha } from "./rede";

describe("rede", () => {
  test('falha de rede vira confiança baixa e a pergunta "descreve em 1 linha"', async () => {
    const item = await estimarComFalha(async () => {
      throw new Error("down");
    });
    expect(item.confianca).toBe("baixa");
    expect(item.pergunta).toBe(PERGUNTA_FALHA);
    expect(item.pergunta).toBe("descreve em 1 linha");
  });

  test("confiança alta omite a pergunta", async () => {
    const item = await estimarComFalha(async () => ({
      kcal: 385,
      p: 18,
      c: 40,
      g: 10,
      confianca: "alto",
      pergunta: "sumir",
    }));
    expect(item.pergunta).toBeNull();
    expect(item.kcal).toBe(385);
  });
});

File: apps/android/app/src/test/java/com/nutri/android/data/RedeTest.kt
package com.nutri.android.data

import app.cash.turbine.test
import com.google.common.truth.Truth.assertThat
import io.mockk.coEvery
import io.mockk.mockk
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.test.runTest
import okhttp3.Request
import org.junit.Test

class RedeTest {
    @Test
    fun `header X-Invite vai na request`() {
        val interceptor = InviteInterceptor("troca-isto")
        val chain = mockk<okhttp3.Interceptor.Chain>()
        val original = Request.Builder().url("http://127.0.0.1:8080/health").build()
        var enviado: Request? = null
        io.mockk.every { chain.request() } returns original
        io.mockk.every { chain.proceed(any()) } answers {
            enviado = firstArg()
            mockk(relaxed = true)
        }
        interceptor.intercept(chain)
        assertThat(enviado!!.header("X-Invite")).isEqualTo("troca-isto")
    }

    @Test
    fun `falha de rede vira confianca baixa e a pergunta de uma linha`() = runTest {
        val api = mockk<NutriApi>()
        coEvery { api.estimate(any()) } throws java.io.IOException("down")
        val gate = EstimateGate(api)
        flow {
            emit(
                gate.estimar(
                    EstimateIn(local_time = "2026-09-24T09:00:00-03:00", janela = "cafe", text = "pao"),
                ),
            )
        }.test {
            val item = awaitItem()
            assertThat(item.confianca).isEqualTo("baixa")
            assertThat(item.pergunta).isEqualTo("descreve em 1 linha")
            awaitComplete()
        }
    }

    @Test
    fun `confianca alta omite a pergunta`() = runTest {
        val api = mockk<NutriApi>()
        coEvery { api.estimate(any()) } returns EstimateOut(kcal = 385.0, p = 18.0, confianca = "alto", pergunta = "sumir")
        val out = EstimateGate(api).estimar(
            EstimateIn(local_time = "2026-09-24T09:00:00-03:00", janela = "cafe", text = "pao"),
        )
        assertThat(out.pergunta).isNull()
        assertThat(out.kcal).isEqualTo(385.0)
    }
}
